File: avalon-support/avalon-oauth2-support/avalon-oauth2-authorization-server/src/main/java/io/github/lishangbu/avalon/oauth2/authorizationserver/autoconfiguration/AuthorizationServerAutoConfiguration.java
package io.github.lishangbu.avalon.oauth2.authorizationserver.autoconfiguration;

import static io.github.lishangbu.avalon.oauth2.common.constant.SecurityBeanDefinitionConstants.AUTHORIZATION_SERVER_SECURITY_FILTER_CHAIN_BEAN_NAME;
import static io.github.lishangbu.avalon.oauth2.common.constant.SecurityBeanDefinitionConstants.AUTHORIZATION_SERVER_SECURITY_FILTER_CHAIN_BEAN_ORDER;

import io.github.lishangbu.avalon.oauth2.authorizationserver.granter.OAuth2PasswordAuthenticationConverter;
import io.github.lishangbu.avalon.oauth2.authorizationserver.granter.OAuth2PasswordAuthenticationProvider;
import io.github.lishangbu.avalon.oauth2.common.properties.Oauth2Properties;
import io.github.lishangbu.avalon.oauth2.common.web.authentication.DefaultAuthenticationEntryPoint;
import java.util.Arrays;
import lombok.RequiredArgsConstructor;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.core.annotation.Order;
import org.springframework.http.MediaType;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.Customizer;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.*;
import org.springframework.security.oauth2.core.OAuth2Token;
import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService;
import org.springframework.security.oauth2.server.authorization.config.annotation.web.configurers.OAuth2AuthorizationServerConfigurer;
import org.springframework.security.oauth2.server.authorization.token.OAuth2TokenGenerator;
import org.springframework.security.oauth2.server.authorization.web.authentication.OAuth2AuthorizationCodeAuthenticationConverter;
import org.springframework.security.oauth2.server.authorization.web.authentication.OAuth2ClientCredentialsAuthenticationConverter;
import org.springframework.security.oauth2.server.authorization.web.authentication.OAuth2RefreshTokenAuthenticationConverter;
import org.springframework.security.web.DefaultSecurityFilterChain;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.DelegatingAuthenticationConverter;
import org.springframework.security.web.util.matcher.MediaTypeRequestMatcher;

/**
 * 自动装配认证服务器
 *
 * @author lishangbu
 * @since 2025/8/17
 */
@EnableWebSecurity
@AutoConfiguration
@EnableMethodSecurity(jsr250Enabled = true, securedEnabled = true)
@RequiredArgsConstructor
public class AuthorizationServerAutoConfiguration {

  private final Oauth2Properties oauth2Properties;

  @Bean
  @Order(AUTHORIZATION_SERVER_SECURITY_FILTER_CHAIN_BEAN_ORDER)
  @ConditionalOnMissingBean(name = AUTHORIZATION_SERVER_SECURITY_FILTER_CHAIN_BEAN_NAME)
  public SecurityFilterChain authorizationServerSecurityFilterChain(HttpSecurity http)
      throws Exception {
    OAuth2AuthorizationServerConfigurer authorizationServerConfigurer =
        OAuth2AuthorizationServerConfigurer.authorizationServer();
    // 禁用csrf和cors
    http.csrf(CsrfConfigurer::disable)
        .cors(CorsConfigurer::disable)
        // 禁用表单登录和会话管理
        .formLogin(FormLoginConfigurer::disable)
        .sessionManagement(SessionManagementConfigurer::disable)
        .rememberMe(RememberMeConfigurer::disable);

    DefaultSecurityFilterChain securityFilterChain =
        http.securityMatcher(authorizationServerConfigurer.getEndpointsMatcher())
            .with(
                authorizationServerConfigurer,
                (authorizationServer) ->
                    // Enable OpenID Connect 1.0
                    authorizationServer
                        .oidc(Customizer.withDefaults())
                        // 在这加上密码模式的转换器
                        .tokenEndpoint(
                            tokenEndpoint ->
                                tokenEndpoint.accessTokenRequestConverter(
                                    new DelegatingAuthenticationConverter(
                                        Arrays.asList(
                                            new OAuth2AuthorizationCodeAuthenticationConverter(),
                                            new OAuth2RefreshTokenAuthenticationConverter(),
                                            new OAuth2ClientCredentialsAuthenticationConverter(),
                                            new OAuth2PasswordAuthenticationConverter())))))
            .authorizeHttpRequests((authorize) -> authorize.anyRequest().authenticated())
            // Redirect to the login page when not authenticated from the
            // authorization endpoint
            .exceptionHandling(
                exceptions -> {
                  exceptions.defaultAuthenticationEntryPointFor(
                      new DefaultAuthenticationEntryPoint(),
                      new MediaTypeRequestMatcher(MediaType.APPLICATION_JSON, MediaType.TEXT_HTML));
                })
            .build();

    addPasswordAuthenticationProvider(http);
    return securityFilterChain;
  }

  @SuppressWarnings("unchecked")
  private void addPasswordAuthenticationProvider(HttpSecurity http) {
    AuthenticationManager authenticationManager = http.getSharedObject(AuthenticationManager.class);
    OAuth2AuthorizationService authorizationService =
        http.getSharedObject(OAuth2AuthorizationService.class);
    OAuth2TokenGenerator<? extends OAuth2Token> tokenGenerator =
        http.getSharedObject(OAuth2TokenGenerator.class);

    OAuth2PasswordAuthenticationProvider resourceOwnerPasswordAuthenticationProvider =
        new OAuth2PasswordAuthenticationProvider(
            authenticationManager, authorizationService, tokenGenerator);

    // This will add new authentication provider in the list of existing authentication providers.
    http.authenticationProvider(resourceOwnerPasswordAuthenticationProvider);
  }
}

File: avalon-support/avalon-oauth2-support/avalon-oauth2-authorization-server/src/main/kotlin/io/github/lishangbu/avalon/oauth2/authorizationserver/autoconfiguration/AuthorizationServerAutoConfiguration.kt
package io.github.lishangbu.avalon.oauth2.authorizationserver.autoconfiguration

import io.github.lishangbu.avalon.oauth2.authorizationserver.login.LoginFailureTracker
import io.github.lishangbu.avalon.oauth2.authorizationserver.web.authentication.AuthorizationEndpointErrorResponseHandler
import io.github.lishangbu.avalon.oauth2.authorizationserver.web.authentication.AuthorizationEndpointResponseHandler
import io.github.lishangbu.avalon.oauth2.authorizationserver.web.authentication.OAuth2AccessTokenApiResultResponseAuthenticationSuccessHandler
import io.github.lishangbu.avalon.oauth2.authorizationserver.web.authentication.OAuth2ErrorApiResultAuthenticationFailureHandler
import io.github.lishangbu.avalon.oauth2.common.constant.SecurityBeanDefinitionConstants.AUTHORIZATION_SERVER_SECURITY_FILTER_CHAIN_BEAN_NAME
import io.github.lishangbu.avalon.oauth2.common.constant.SecurityBeanDefinitionConstants.AUTHORIZATION_SERVER_SECURITY_FILTER_CHAIN_BEAN_ORDER
import io.github.lishangbu.avalon.oauth2.common.properties.Oauth2Properties
import io.github.lishangbu.avalon.oauth2.common.web.authentication.DefaultAuthenticationEntryPoint
import org.springframework.boot.autoconfigure.AutoConfiguration
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.context.annotation.Bean
import org.springframework.core.annotation.Order
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.config.Customizer
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configurers.*
import org.springframework.security.config.annotation.web.configurers.oauth2.server.authorization.OAuth2AuthorizationServerConfigurer
import org.springframework.security.oauth2.core.OAuth2Token
import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService
import org.springframework.security.oauth2.server.authorization.authentication.OAuth2EmailAuthenticationProvider
import org.springframework.security.oauth2.server.authorization.authentication.OAuth2PasswordAuthenticationProvider
import org.springframework.security.oauth2.server.authorization.authentication.OAuth2SmsAuthenticationProvider
import org.springframework.security.oauth2.server.authorization.token.OAuth2TokenGenerator
import org.springframework.security.oauth2.server.authorization.web.authentication.*
import org.springframework.security.web.DefaultSecurityFilterChain
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.DelegatingAuthenticationConverter
import tools.jackson.databind.json.JsonMapper

/**
 * 授权服务器自动配置
 *
 * 提供授权端点、令牌端点和扩展授权类型所需的安全过滤链
 */
@EnableWebSecurity
@AutoConfiguration
@EnableMethodSecurity(jsr250Enabled = true, securedEnabled = true)
class AuthorizationServerAutoConfiguration {
    /**
     * 创建授权服务器安全过滤链
     */
    @Bean
    @Order(AUTHORIZATION_SERVER_SECURITY_FILTER_CHAIN_BEAN_ORDER)
    @ConditionalOnMissingBean(name = [AUTHORIZATION_SERVER_SECURITY_FILTER_CHAIN_BEAN_NAME])
    @Throws(Exception::class)
    fun authorizationServerSecurityFilterChain(
        http: HttpSecurity,
        oauth2Properties: Oauth2Properties,
        accessTokenResponseAuthenticationSuccessHandler: OAuth2AccessTokenApiResultResponseAuthenticationSuccessHandler,
        oauth2ErrorApiResultAuthenticationFailureHandler: OAuth2ErrorApiResultAuthenticationFailureHandler,
        loginFailureTracker: LoginFailureTracker,
        jsonMapper: JsonMapper,
    ): SecurityFilterChain {
        val authorizationServerConfigurer = OAuth2AuthorizationServerConfigurer()

        // 禁用 CSRF 和 CORS
        http
            .csrf(CsrfConfigurer<HttpSecurity>::disable)
            .cors(CorsConfigurer<HttpSecurity>::disable)
            // 禁用表单登录、会话管理和记住我
            .formLogin(FormLoginConfigurer<HttpSecurity>::disable)
            .sessionManagement(SessionManagementConfigurer<HttpSecurity>::disable)
            .rememberMe(RememberMeConfigurer<HttpSecurity>::disable)

        val securityFilterChain: DefaultSecurityFilterChain =
            http
                .securityMatcher(authorizationServerConfigurer.endpointsMatcher)
                .with(authorizationServerConfigurer) { authorizationServer ->
                    // 启用 OpenID Connect 1.0
                    authorizationServer
                        .authorizationEndpoint { authorizationEndpoint ->
                            authorizationEndpoint
                                .authorizationResponseHandler(
                                    AuthorizationEndpointResponseHandler(),
                                ).errorResponseHandler(
                                    AuthorizationEndpointErrorResponseHandler(jsonMapper),
                                )
                        }.oidc(Customizer.withDefaults())
                        // 统一客户端认证失败响应
                        .clientAuthentication { clientAuthentication ->
                            clientAuthentication.errorResponseHandler(
                                oauth2ErrorApiResultAuthenticationFailureHandler,
                            )
                        }.tokenEndpoint { tokenEndpoint ->
                            tokenEndpoint
                                // 定制成功响应
                                .accessTokenResponseHandler(
                                    accessTokenResponseAuthenticationSuccessHandler,
                                )
                                // 定制失败响应
                                .errorResponseHandler(
                                    oauth2ErrorApiResultAuthenticationFailureHandler,
                                )
                                // 注册扩展授权类型转换器
                                .accessTokenRequestConverter(
                                    DelegatingAuthenticationConverter(
                                        listOf(
                                            OAuth2AuthorizationCodeAuthenticationConverter(),
                                            OAuth2RefreshTokenAuthenticationConverter(),
                                            OAuth2ClientCredentialsAuthenticationConverter(),
                                            OAuth2PasswordAuthenticationConverter(oauth2Properties),
                                            OAuth2SmsAuthenticationConverter(oauth2Properties),
                                            OAuth2EmailAuthenticationConverter(oauth2Properties),
                                        ),
                                    ),
                                )
                        }
                }.authorizeHttpRequests { authorize -> authorize.anyRequest().authenticated() }
                .oauth2Login { oauth2Login ->
                    oauth2Login
                        .successHandler(AuthorizationEndpointResponseHandler())
                        .failureHandler(AuthorizationEndpointErrorResponseHandler(jsonMapper))
                }.exceptionHandling { exceptions ->
                    // 使用统一的 AuthenticationEntryPoint，避免 MediaTypeRequestMatcher
                    // 导致的不匹配
                    exceptions.authenticationEntryPoint(DefaultAuthenticationEntryPoint(jsonMapper))
                }.build()

        addAdditionalAuthenticationProvider(http, loginFailureTracker)
        return securityFilterChain
    }

    /**
     * 注册扩展授权类型的认证提供者
     */
    @Suppress("UNCHECKED_CAST")
    private fun addAdditionalAuthenticationProvider(
        http: HttpSecurity,
        loginFailureTracker: LoginFailureTracker,
    ) {
        val authenticationManager = http.getSharedObject(AuthenticationManager::class.java)
        val authorizationService = http.getSharedObject(OAuth2AuthorizationService::class.java)
        val tokenGenerator =
            http.getSharedObject(OAuth2TokenGenerator::class.java)
                as OAuth2TokenGenerator<out OAuth2Token>

        val resourceOwnerPasswordAuthenticationProvider =
            OAuth2PasswordAuthenticationProvider(
                authenticationManager,
                authorizationService,
                tokenGenerator,
                loginFailureTracker,
            )

        val smsAuthenticationProvider =
            OAuth2SmsAuthenticationProvider(
                authenticationManager,
                authorizationService,
                tokenGenerator,
                loginFailureTracker,
            )

        val emailAuthenticationProvider =
            OAuth2EmailAuthenticationProvider(
                authenticationManager,
                authorizationService,
                tokenGenerator,
                loginFailureTracker,
            )

        // 将扩展认证提供者追加到现有认证链
        http.authenticationProvider(resourceOwnerPasswordAuthenticationProvider)
        http.authenticationProvider(smsAuthenticationProvider)
        http.authenticationProvider(emailAuthenticationProvider)
    }
}
